### PR TITLE
[x86-64] Slightly improved Sign function (and new conformance tests)

### PR DESCRIPTION
**Pull request by CuriousKit**
This pull request updates the Sign function to use a slightly different algorithm under x86-64 that should run about one or two cycles faster thanks to parallel execution and reducing the dependency chain.

Additionally, three new conformance tests have been added to TestGR32_LowLevel to evaluate the correctness of the Sign function for 0 (the only time Sign returns 0) and the extreme values $7FFFFFFF (largest positive) and $80000000 (largest negative). @andersmelander, can you confirm correctness and timing (even though the cycle count is very low already)?

### CODE REVIEW

File: Tests/LowLevel Test/TestGR32_LowLevel.pas [hunk 1]
```diff
@@ -1341,6 +1341,9 @@ procedure TTestMath.TestSign;
 var
   Index : Integer;
 begin
+  CheckEquals(0, GR32_Math.Sign(0));
+  CheckEquals(1, GR32_Math.Sign($7FFFFFFF));
+  CheckEquals(-1, GR32_Math.Sign(Integer($80000000)));
```
Comment (andersmelander) [inline Tests/LowLevel Test/TestGR32_LowLevel.pas:1346]: The PUREPASCAL version fails for the test case `Sign($80000000) = -1`.
However, that is an existing problem which your added tests has surfaced, so just FYI.

File: Source/GR32_Math.pas [hunk 2]
```diff
@@ -1224,13 +1224,26 @@ function Sign(Value: Integer): Integer;
 {$IFDEF FPC} assembler; nostackframe; {$ENDIF}
 asm
 {$IFDEF TARGET_x64}
-        MOV     EAX, Value
-{$ENDIF}
+  {$IFDEF MSWINDOWS}
```
Comment (andersmelander) [inline Source/GR32_Math.pas:1227]: Looks good - But why not use the same for 32-bit?

File: Source/GR32_Math.pas [hunk 3]
```diff
@@ -1224,13 +1224,26 @@ function Sign(Value: Integer): Integer;
...
 {$IFDEF TARGET_x64}
-        MOV     EAX, Value
-{$ENDIF}
+  {$IFDEF MSWINDOWS}
+        XOR     EDX, EDX
+        TEST    ECX, ECX
+        SETG    DL
+        SAR     ECX, 31
+        LEA     EAX, [EDX + ECX]
+  {$ELSE}
```
Comment (andersmelander) [inline Source/GR32_Math.pas:1233]: You need to select the code block based on some criteria instead of just assuming that everything not Windows can use that calling convention.
I assume that FPC has a symbol defined for your target that can be used.

So somthing like:
```Delphi
{$if defined(Windows)}
        XOR     EDX, EDX
        TEST    ECX, ECX
        [...]
{$elseif defined(YourPlatform)}
        XOR     EDX, EDX
        TEST    EDI, EDI
        [...]
{ifend}
```

**Review by andersmelander — CHANGES_REQUESTED** [review]

File: Tests/LowLevel Test/TestGR32_LowLevel.pas (hunk 1 above)
Comment (CuriousKit) [reply Tests/LowLevel Test/TestGR32_LowLevel.pas:1346]: Does Delphi have an equivalent to Free Pascal's "SarLongint" intrinsic?  That would permit the use of arithmetic shifts.  If not, it might have to be an explicit check for $80000000, since negating that value causes an overflow.

File: Source/GR32_Math.pas (hunk 3 above)
Comment (CuriousKit) [reply Source/GR32_Math.pas:1233]: As far as I know, all non-Windows platforms use the "System V ABI", whether Linux or MacOSX etc; only Microsoft chose to do something different.

Selecting the code block after checking to see if it's Linux or Darwin... I'm bound to miss a platform!  Or do you insist it be done this way?

File: Source/GR32_Math.pas (hunk 2 above)
Comment (CuriousKit) [reply Source/GR32_Math.pas:1227]: This was due to careful cycle counting.  Under 32-bit, the cycle count is about 3, assuming NEG can execute at the same time as CDQ.  Under 64-bit, the cycle count is at least 4 due to the need to move the formal parameter into EAX first.

For the new version, XOR and TEST can execute at the same time (1 cycle apiece) as they work on different registers, while SETG depends on both of these instructions completing, while SAR ECX, 31 can execute at the same time, before the results are merged into the LEA instruction for 3 cycles overall.  From my own experimentation in the past, if you zero the whole register, there's no partial write penalty from using SETcc on the mapped byte register.

Long story short, the reason why the 32-bit vrersion wasn't changed is because there's no actual speed saving, but the existing algorithm takes fewer bytes of machine code at only 5 bytes (ultimately it probably won't matter since code pointers are aligned to at least a 16-byte granularity).

File: Tests/LowLevel Test/TestGR32_LowLevel.pas (hunk 1 above)
Comment (andersmelander) [reply Tests/LowLevel Test/TestGR32_LowLevel.pas:1346]: Not really. There is only `x div 2^n` which is compiled to SAR on 64-bit but unfortunately:
https://github.com/graphics32/graphics32/blob/8265594faf795163da9ea32a0bcd559605876c35/Source/GR32_LowLevel.pas#L252-L267

I think that if we have to test for a specific value then that will outweigh the benefit of the trick applied. Better to just document it.

Comment (CuriousKit) [reply Tests/LowLevel Test/TestGR32_LowLevel.pas:1346]: I'll have a think on that one.  The Sign function behaving differently between assembly and Pascal implementations, even if just for one extreme value, is not ideal.

File: Source/GR32_Math.pas (hunk 3 above)
Comment (CuriousKit) [reply Source/GR32_Math.pas:1233]: If necessary, a compilation error can be forced with {$ERROR ... } if an architecture/platform isn't fully defined, or just simply fall back to pure Pascal.  At a future date it might be plausible to separate the assembly implementations into separate .inc files.

File: Source/GR32_Math.pas (hunk 2 above)
Comment (andersmelander) [reply Source/GR32_Math.pas:1227]: > Under 64-bit, the cycle count is at least 4 due to the need to move the formal parameter into EAX first.

What move? You mean the LEA?
Apart from that I got the same timings.

| Op | Latency |
| ---- | ---- |
| XOR | 0/1 |
| TEST | 1 |
| SETG | 1 |
| SAR | 1 | 
| LEA |  1 |

| Op | Latency |
| ---- | ---- |
| CDQ | 1 |
| NEG | 1 |
| ADC | 1 |
| MOV | 1 |

File: Tests/LowLevel Test/TestGR32_LowLevel.pas (hunk 1 above)
Comment (andersmelander) [reply Tests/LowLevel Test/TestGR32_LowLevel.pas:1346]: We could just defer to the RTL Sign function and let inlining sort the overhead out.
That is already done in a few other places where a faster PUREPASCAL version isn't possible.

Comment (andersmelander) [reply Tests/LowLevel Test/TestGR32_LowLevel.pas:1346]: >  behaving differently between assembly and Pascal implementations, even if just for one extreme value, is not ideal.

You are going to really enjoy the blend/merge functions then :-)

File: Source/GR32_Math.pas (hunk 2 above)
Comment (CuriousKit) [reply Source/GR32_Math.pas:1227]: I meant the ```MOV EAX, Value``` instruction that has to appear before ```CDQ``` on 64-bit, which will add at least 1 to the latency.

File: Tests/LowLevel Test/TestGR32_LowLevel.pas (hunk 1 above)
Comment (CuriousKit) [reply Tests/LowLevel Test/TestGR32_LowLevel.pas:1355]: > We could just defer to the RTL Sign function and let inlining sort the overhead out. That is already done in a few other places where a faster PUREPASCAL version isn't possible.

I've pushed an extra commit that defers to Math.Sign under PUREPASCAL.  Does this work for you?

File: Source/GR32_Math.pas (hunk 2 above)
Comment (CuriousKit) [reply Source/GR32_Math.pas:1229]: In the meantime I added an extra comment for the 32-bit version to explain why it's still using the old algorithm.
